tests/feature_resolution_editions.js

```diff
@@ -0,0 +1,89 @@
+var tape = require("tape");
+
+var protobuf = require("..");
+
+var protoEditions2023 = `edition = "2023";`;
+
+var proto2 = `syntax = "proto2";`;
+
+var proto3 = `syntax = "proto3";`;
+
+var editions2023Defaults = {enum_type: 'OPEN', field_presence: 'EXPLICIT', json_format: 'ALLOW', message_encoding: 'LENGTH_PREFIXED', repeated_field_encoding: 'PACKED', utf8_validation: 'VERIFY'}
+var proto2Defaults = {enum_type: 'CLOSED', field_presence: 'EXPLICIT', json_format: 'LEGACY_BEST_EFFORT', message_encoding: 'LENGTH_PREFIXED', repeated_field_encoding: 'EXPANDED', utf8_validation: 'NONE'}
+var proto3Defaults = {enum_type: 'OPEN', field_presence: 'IMPLICIT', json_format: 'ALLOW', message_encoding: 'LENGTH_PREFIXED', repeated_field_encoding: 'PACKED', utf8_validation: 'VERIFY'}
+
+
+var protoEditions2023Overridden = `edition = "2023";
+option features.json_format = LEGACY_BEST_EFFORT;
+
+option features.(abc).d_e = deeply_nested_false;
+
+message Message {
+    string string_val = 1;
+    string string_repeated = 2 [features.enum_type = CLOSED];
+
+    message Nested {
+        option features.(abc).d_e = deeply_nested_true;
+        option features.field_presence = IMPLICIT;
+        int64 count = 9;
+    }
+}
+`
+tape.test("feautre resolution defaults", function(test) {
+    var rootEditions = protobuf.parse(protoEditions2023).root;
+    test.same(rootEditions._features, editions2023Defaults);
+
+    var rootProto2 = protobuf.parse(proto2).root;
+    test.same(rootProto2._features, proto2Defaults);
+
+    var rootProto3 = protobuf.parse(proto3).root;
+    test.same(rootProto3._features, proto3Defaults);
+
+    test.end();
+})
+
+tape.test("feature resolution inheritance", function(test) {
+    var rootEditionsOverriden = protobuf.parse(protoEditions2023Overridden).root
+
+    rootEditionsOverriden.resolveAll();
+
+    // Should flip enum_type from default setting, inherit from Message,
+    // and keep everything else
+    test.same(rootEditionsOverriden.lookup("Message").fields.stringRepeated._features, {
+        enum_type: 'CLOSED',
+        field_presence: 'EXPLICIT',
+        json_format: 'LEGACY_BEST_EFFORT',
+        message_encoding: 'LENGTH_PREFIXED',
+        repeated_field_encoding: 'PACKED',
+        utf8_validation: 'VERIFY',
+        '(abc)': { d_e: 'deeply_nested_false' }
+      })
+
+    // Should inherit from default, and Message, only change field_presence and the custom extension
+    test.same(rootEditionsOverriden.lookup("Message").lookup("Nested")._features, 
+    { enum_type: 'OPEN', field_presence: 'IMPLICIT', json_format: 'LEGACY_BEST_EFFORT', message_encoding: 'LENGTH_PREFIXED', repeated_field_encoding: 'PACKED', utf8_validation: 'VERIFY', '(abc)': { d_e: 'deeply_nested_true' } })
+
+    test.end();
+})
+// Tests precedence for different levels of feature resolution
+tape.test("feature resolution editions precedence", function(test) {
+
+    protobuf.load("tests/data/feature-resolution.proto", function(err, root) {
```

Comment:
I think it might be helpful to split up this proto into smaller ones, and define them inline per-test.  I regret using monolithic protos in our tests, as it's much harder to understand what's being tests

Reply:
I hear you, but I think in this case it helps the test reader understand the overall precedence of options - at least it does for me. It's clear by the alphabetical order what has precedence in the grand scheme of things, rather than having to piece it together in each test. As long as it doesn't get used for other development, I think it's helpful to get the greater overall picture of precendence.

Reply:
The problem is that to understand this test you need to constantly be switching back and forth between two different files (as a reader, I find this test difficult to read).  If you had much smaller protos defined inline in each test, it would be much clearer what the expectation is

Reply:
OK, switched it.